[WIP] Merge library rebnk to master

Granted07 opened the pull request:
## Goals of rebnk:
- [x] Read the header of BNK files
- [x] Read the data index of BNK files
- [x] Extract raw WEM streams
- [ ] Identify audio types from HIRC section
- [ ] Add feature to group WEM streams from HIRC playlist data and generate non truncated WEM
- [ ] [Optional] Convert from WEM to OGG and opposite
- [ ] Build .bnk files from WEM data

Granted07 committed:
innit rebnk

Granted07 committed:
reading header files

natimerry committed:
refactor: use sans I/O implementation as library code doesnt rely on IO implementation

natimerry commented:
```
git fetch origin Granted07/master
git merge origin/Granted07/master
```

Merge my changes before working on this further

Granted07 committed:
remove need to create buffer to parse BnkHeader and pass in file handle to read

remove io

remove io

natimerry committed:
refactor: create WwiseReader struct which stores the offsets to the other sections

This allows us to create functions like `read_index` or `write_index` which will only read the data when required. Internally for the `data` section we can return a Map of WEMS maybe. Consult `footer` code in repak for design guide

natimerry commented on rebnk/src/wwise_reader.rs:
```diff
@@ -0,0 +1,59 @@
+                    hirc_offset = current_offset;
+                }
+                _ => {
+                    // Right now we just continue since DIDX data and HIRC hasnt gotten implemented yet
```
This logic is still so stupid even after we are done parsing since there can be arbitrary padding bytes or other stuff. Maybe keep this flexible in some different way?

Granted07 committed:
added functionality for reading didx section

natimerry committed:
add `wem_length` parameter to WemEntry while making renaming changes

natimerry committed:
refactor: use a hashmap for entries to index by ID

natimerry committed:
feature: extract wem files

natimerry committed:
add stub hirc and sound structure data
